### PR TITLE
Implementation of BIN$ & HEX$ functions

### PR DESCRIPTION
**Pull request by JimmyDansbo**
An implementation of the BIN$ and HEX$ functions as suggested in issue #49 
Both functions only work with byte input because of space restraints in the 16KB BASIC ROM page.
The hexd function does a jsrfar to the byte_to_hex_ascii function in monitor.s, but the address is hard coded as symbols are not shared between BASIC and MONITOR pages.

### CODE REVIEW

File: inc/mac.inc [hunk 1]
```diff
@@ -20,7 +20,7 @@
 		lda #<(value)
 		sta dest
 	.endif
-	.if .const(value) .and value=0
+	.if (.const(value) .and .hibyte(value)=0)
```
Comment (greg-king5) [inline inc/mac.inc:23]: You can use that same improvement for the low byte of `LoadW`.

Comment (JimmyDansbo) [reply inc/mac.inc:23]: You are right, I missed that :-)

File: basic/x16additions.s [hunk 2]
```diff
@@ -90,6 +90,71 @@ coltab	;this is an unavoidable duplicate from KERNAL
...
+; lofbuf+1 = $0100 so .X can be used as index
+; without trying to cross page
+	txa
+	ldx #7
+@loop:	lsr        ; low bit to carry
+	tay	   ; save .A for next iteration
+	lda #'1'
+	bcs :+
+	dec
+:	sta lofbuf+1,x
```
Comment (mist64) [inline basic/x16additions.s:111]: you can use
   sta a:lobuf,x
to force
  STA $00FF,X
which will not wrap around in zero page.

**Comment by mist64** [comment]
Sorry it took to song for me to look at this.

I like the feature, and I like the code, but I really think we should support 16 bit values, like on the C128.

By the way, we can legally use the C128 source:

https://github.com/mist64/cbmsrc/blob/master/BASIC_C128/hexfunc.src

**Comment by JimmyDansbo** [comment]
Nice to see you back, Michael :)
I am not sure when I will have time to look into this further. Should I just close this pull request?
Hopefully someone else can help with implementing 16bit version of these functions.

**Comment by mist64** [comment]
Leave it open. I'll think of something.

File: basic/x16additions.s (hunk 2 above)
Comment (JimmyDansbo) [reply basic/x16additions.s:111]: I was not aware of the ability to force the address size of a symbol in ca65, but found it in the documentation here: https://cc65.github.io/doc/ca65.html#ss8.3 - Thanks.

File: inc/mac.inc [hunk 3]
```diff
@@ -14,13 +14,13 @@
 .endmacro
 
 .macro LoadW dest, value
-	.if .const(value) .and value=0
+	.if .const(value) .and .lobyte(value)=0
```
Comment (mist64) [inline inc/mac.inc:17]: 👍

File: basic/x16additions.s (hunk 2 above)
Comment (JimmyDansbo) [reply basic/x16additions.s]: Now fixed

File: inc/mac.inc (hunk 3 above)
Comment (JimmyDansbo) [reply inc/mac.inc:17]: :smile:

**Comment by JimmyDansbo** [comment]
@mist64 , would you prefer the HEX$ and BIN$ functions to precede the returned string with '$' and '%' respectively? At the moment they are returning strings with varying length depending on the number of bits being passed to them (8 or 16).

**Comment by mist64** [comment]
The C128 always returns 4 digits for HEX$() without a prefix and does not support BIN$(). I'm tempted to say go with C128 compatibility...

But... I won't say that. Variable length is better. But no prefix.

**Comment by JimmyDansbo** [comment]
Great, as far as I can see, that should be the current functionality.
Unless I have f*cked something up, I hope this can be merged?

**Comment by mist64** [comment]
:)

**Comment by irmen** [comment]
awesome to see this merged with support for words

**Pull request merged**